firefox: start killing all the per-extension options

K900 opened the pull request:
## Description of changes

Deprecate all (incomplete, partially overlapping) extension-specific options in the wrapper and the module, to hopefully kill them off in 24.05.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

K900 committed:
firefox: start killing all the per-extension options

SuperSandro2000 commented on nixos/modules/programs/firefox.nix:
```diff
@@ -204,46 +246,35 @@ in
-      tridactyl = "Tridactyl support";
-      ugetIntegrator = "Uget Integrator support";
-    };
+    nativeMessagingHosts = ({
+      packages = mkOption {
```
Stupid question: what else could we put here that justifies adding packages? Otherwise we could just name it ``nativeMessagingHostPackages`` and/or rename it later when it is actually required.

K900 replied:
All the existing options are already under there, so we can't get rid of the namespace (yet).

K900 replied:
I'd like to eventually just get rid of all the options and make `nativeMessagingHosts` a list.

K900 replied:
We'll have a hard break when the deprecated options are removed.

SuperSandro2000 replied:
until then users are informed with a warning

SuperSandro2000 commented on nixos/modules/programs/firefox.nix:
```diff
@@ -204,46 +246,35 @@ in
+  config = let
+    forEachEnabledNmh = fn: flatten (mapAttrsToList (lib.optional cfg.nativeMessagingHosts.${k} (fn k v)) fn nmhOptions);
+  in mkIf cfg.enable {
+    warnings = forEachEnabledNmh (k: v: "The `programs.firefox.nativeMessagingHosts.${k}` option is deprecated, please add `${v.package.pname}` to `programs.firefox.nativeMessagingHosts.packages` instead.");
```
```suggestion
    warnings = forEachEnabledNmh (k: v:
      "The `programs.firefox.nativeMessagingHosts.${k}` option is deprecated, please add `${v.package.pname}` to `programs.firefox.nativeMessagingHosts.packages` instead.");
```

K900 replied:
Added some line breaks.

SuperSandro2000 reviewed (approved):
like the idea a lot making this better extensible

K900 commented:
The worst part is that it was already extensible, just a bunch of weird layering all over the place.

aij commented:
I gave this a quick try with `programs.firefox.enable` and `programs.browserpass.enable`, but I'm getting:

```console
ivan@tobati:~/nixops/aij$ nixos-rebuild build
building Nix...
building the system configuration...
error:
       … while calling the 'head' builtin

         at /home/ivan/nixops/aij/unstable/lib/attrsets.nix:850:11:

          849|         || pred here (elemAt values 1) (head values) then
          850|           head values
             |           ^
          851|         else

       … while evaluating the attribute 'value'

         at /home/ivan/nixops/aij/unstable/lib/modules.nix:807:9:

          806|     in warnDeprecation opt //
          807|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          808|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: value is a function while a set was expected
```

K900 committed:
nixos/tests/firefox: use the module

This should exercise the module a bit in the test, and also removes an awkward override.

K900 commented:
Please try again.

danth commented on nixos/modules/programs/firefox.nix:
```diff
@@ -204,46 +246,31 @@ in
+      packages = mkOption {
+        type = types.listOf types.package;
+        default = [];
+        description = mdDoc ''
+          Additional packages containing native messaging hosts that should be made available to Firefox extensions.
+        '';
```
Could we give a suggestion of which packages to use here? Similar to `relatedPackages` for the `programs.firefox.package` option

K900 replied:
I really don't want to have another list of all the extensions here...

danth replied:
Fair point, there would be a lot of packages if we listed them all.

aij reviewed (approved):
I tested again with 60421a1622ac05433afcf0179ff46f937e10096c and now browserpass does work without needing to set  the old `nixpkgs.config.firefox.enableBrowserpass` option any more. Thanks!

I see you updated the test to use `programs.firefox.enable` too now. :+1: 

Do we need to update release notes to let people know they may need to switch to `programs.firefox.enable` instead of installing the `firefox` package?

herberteuler commented:
`nativeMessagingHosts` currently only exists in FireFox's NixOS module. Is there a plan to make it available for `nix-env`?

K900 commented:
You can pass it with an override, which is what the module already does internally.

herberteuler commented:
> You can pass it with an override, which is what the module already does internally.

Thanks for the quick reply. I deleted my comment because I was dubious of the practicability of installing FireFox using `nix-env` immediately after posting it. On a second thought, I thought it might be more sensible to install FireFox on host platforms while using `nix-env`.

For the record for whoever is interested in my comment, it was "`nativeMessagingHosts` currently only exists in FireFox's NixOS module. Is there a plan to make it available for `nix-env`?"
